rts/motoko-rts/src/gc/copying.rs

```diff
@@ -0,0 +1,158 @@
+///
+/// Arguments:
+///
+/// - begin_from_space: Where the dynamic heap starts. Used for two things:
+///
+///   - An object is static if its address is below this value. These objects either don't point to
+///     dynamic heap, or listed in static_roots array. Objects in static_roots are scavenged
+///     separately in `evac_static_roots` below. So we skip these objects here.
+///
+///   - After all objects are evacuated we move to-space to from-space, to be able to do that the
+///     pointers need to point to their (eventual) locations in from-space, which is calculated with
+///     `end_to_space - begin_to_space + begin_from_space`.
+///
+/// - begin_to_space: Where to-space starts. See above for how this is used.
+///
+/// - end_to_space: Where the object in `ptr_loc` will be copied.
+///
+/// - ptr_loc: Location of the object to evacuate, e.g. an object field address.
+///
+unsafe fn evac(
+    begin_from_space: usize,
+    begin_to_space: usize,
+    end_to_space: &mut usize,
+    ptr_loc: usize,
+) {
+    // Field holds a skewed pointer to the object to evacuate
+    let ptr_loc = ptr_loc as *mut SkewedPtr;
+
+    let obj = (*ptr_loc).unskew() as *mut Obj;
+
+    // Update the field if the object is already evacauted
+    if obj.tag() == TAG_FWD_PTR {
+        let fwd = (*(obj as *const FwdPtr)).fwd;
+        *ptr_loc = fwd;
+        return;
+    }
+
+    let obj_size = object_size(obj as usize);
+    let obj_size_bytes = obj_size.to_bytes();
+
+    // Grow memory if needed
+    alloc::grow_memory(*end_to_space + obj_size_bytes.0 as usize);
+
+    // Copy object to to-space
+    memcpy_words(*end_to_space, obj as usize, obj_size);
+
+    // Final location of the object after copying to-space back to from-space
+    let obj_loc = (*end_to_space - begin_to_space) + begin_from_space;
+
+    // Set forwarding pointer
+    let fwd = obj as *mut FwdPtr;
+    (*fwd).header.tag = TAG_FWD_PTR;
+    (*fwd).fwd = skew(obj_loc);
+
+    // Update evacuated field
+    *ptr_loc = skew(obj_loc);
+
+    // Update end of to-space
+    *end_to_space += obj_size_bytes.0 as usize
+}
+
+unsafe fn scav(
+    begin_from_space: usize,
+    begin_to_space: usize,
+    end_to_space: &mut usize,
+    obj: usize,
+) {
+    let obj = obj as *mut Obj;
+
+    crate::visitor::visit_pointer_fields(obj, begin_from_space, |field_addr| {
+        evac(
+            begin_from_space,
+            begin_to_space,
+            end_to_space,
+            field_addr as usize,
+        );
+    });
+}
+
+// We have a special evacuation routine for "static roots" array: we don't evacuate elements of
+// "static roots", we just scavenge them.
+unsafe fn evac_static_roots(
+    begin_from_space: usize,
+    begin_to_space: usize,
+    end_to_space: &mut usize,
+    roots: *mut Array,
+) {
+    // The array and the objects pointed by the array are all static so we don't evacuate them. We
+    // only evacuate fields of objects in the array.
+    for i in 0..roots.len() {
```

Comment:
Do we actually know that all the values in roots are pointers to heap objects and never unboxed value or even static objects? Otherwise this code might crash, right?

Reply:
My reasoning is that env_static_roots eventually calls visit_pointer_fields, but there's not check that the obj is a proper object.

Reply:
Yeah, so the static heap (other than static data/constants) have a an array of `MutBox`es (`roots` in this function). These `MutBox`es also live in static heap but they can point to dynamic heap. This code evacuates those `MutBox` fields.

Reply:
Right, just remembered that after (re)seeing the explanation to a previous comment. Sorry about the gaslighting